crates/iroha_torii/src/stream.rs

```diff
@@ -17,124 +15,97 @@ const TIMEOUT: Duration = Duration::from_millis(1000);
 #[derive(Debug, displaydoc::Display, thiserror::Error)]
 #[ignore_extra_doc_attributes]
-pub enum Error<InternalStreamError>
-where
-    InternalStreamError: std::error::Error + Send + Sync + 'static,
-{
+pub enum Error {
     /// Read message timeout
     ReadTimeout,
     /// Send message timeout
     SendTimeout,
-    /// An empty message was received
-    NoMessage,
-    /// Error in internal stream representation (typically `WebSocket`)
-    ///
-    /// Made without `from` macro because it will break `IrohaVersion` variant conversion
-    InternalStream(#[source] InternalStreamError),
-    /// `Close` message received
-    CloseMessage,
-    /// Unexpected non-binary message received
-    NonBinaryMessage,
+    /// WebSocket error: {_0}
+    WebSocket(#[source] axum::Error),
     /// Error during versioned message decoding
     Decode(#[from] parity_scale_codec::Error),
+    /// Connection is closed
+    Closed,
 }
 
-/// Represents message used by the stream
-pub trait StreamMessage {
-    /// Construct new binary message
-    fn binary(source: Vec<u8>) -> Self;
+/// Wrapper to send/receive scale encoded messages
+#[derive(Debug)]
+pub struct WebSocketScale(pub(crate) WebSocket);
 
-    /// Check if message is binary and if so return payload
-    fn try_binary(self) -> Option<Vec<u8>>;
-
-    /// Returns `true` if it's a closing message
-    fn is_close(&self) -> bool;
-}
-
-/// Trait for writing custom messages into stream
-#[async_trait::async_trait]
-pub trait Sink<S>: SinkExt<Self::Message, Error = Self::Err> + Unpin
-where
-    S: Encode + Send + Sync + 'static,
-{
-    /// Error type returned by the sink
-    type Err: std::error::Error + Send + Sync + 'static;
-
-    /// Message type used by the underlying sink
-    type Message: StreamMessage + Send;
-
-    /// Encoded message and sends it to the stream
-    async fn send(&mut self, message: S) -> Result<(), Error<Self::Err>> {
-        tokio::time::timeout(
-            TIMEOUT,
-            <Self as SinkExt<Self::Message>>::send(self, Self::Message::binary(message.encode())),
-        )
-        .await
-        .map_err(|_err| Error::SendTimeout)?
-        .map_err(Error::InternalStream)
-    }
-}
-
-/// Trait for reading custom messages from stream
-#[async_trait::async_trait]
-pub trait Stream<R: DecodeAll>:
-    StreamExt<Item = std::result::Result<Self::Message, Self::Err>> + Unpin
-{
-    /// Error type returned by the stream
-    type Err: std::error::Error + Send + Sync + 'static;
-
-    /// Message type used by the underlying stream
-    type Message: StreamMessage;
-
-    /// Receives and decodes message from the stream
-    async fn recv(&mut self) -> Result<R, Error<Self::Err>> {
-        let subscription_request_message = tokio::time::timeout(TIMEOUT, self.next())
+impl WebSocketScale {
+    /// Send message encoded in scale
+    pub async fn send<M: Encode + Send>(&mut self, message: M) -> Result<(), Error> {
```

Comment:
Would it make sense to lift these generics (along with the ones on `recv`) to the `WebSocketScale` type? I don't think we expect to receive different types of messages during the lifetime of the connection.

Something like `struct WebSocketScale<ReceivedMessage, SentMessage>`

Reply:
I think placing generic on `recv` and `send` allow implementing handshake procedure:
1. recv/send handshake msg type
2. exchange ordinary messages